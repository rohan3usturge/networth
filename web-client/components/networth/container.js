import { CurrencyForm } from "./../currency/currency-form";
import { AssetCategoryList } from "./asset-category-list";
import { LiabilityCategoryList } from "./liability-category-list";
import { useState } from "react";
import { NETWORTH_API } from "../../apis/networth_api";
import { NUMBER_UTILS } from "../utils";
import { withTranslation } from "react-i18next";
import { useRouter } from "next/router";

const Wrapper = ({ defaultPortfolio, currencies, t }) => {
  const [portfolio, setPortfolio] = useState(defaultPortfolio || {});
  const [isLoading, setIsLoading] = useState(false);

  function timeout(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  const fetchNetWorth = async (currentPortfolio, targetCurrencyCode) => {
    setIsLoading(true);
    targetCurrencyCode = targetCurrencyCode || currentPortfolio.currencyCode;
    const nwRequest = {
      portfolio: currentPortfolio,
      targetCurrencyCode,
    };
    const nw = await NETWORTH_API.getNetWorth(nwRequest);
    await timeout(300);
    setIsLoading(false);
    setPortfolio(nw.portfolio);
  };

  const handleLiabilitiesChange = (changedL) => {
    const changed = { ...portfolio };
    changed.lineItems.liabilities = changedL;
    setPortfolio(changed);
    fetchNetWorth(changed);
  };

  const handleAssetChange = (changedA) => {
    const changed = { ...portfolio };
    changed.lineItems.assets = changedA;
    setPortfolio(changed);
    fetchNetWorth(changed);
  };

  const handleCurrencyChange = (changedC) => {
    fetchNetWorth(portfolio, changedC);
  };

  return (
    <>
      {isLoading && (
        <div className="d-flex justify-content-center">
          <div className="spinner-grow text-primary" role="status">
            <span className="sr-only">Loading...</span>
          </div>
        </div>
      )}

      <CurrencyForm
        selectedCurrency={portfolio.currencyCode}
        currencies={currencies}
        onChange={handleCurrencyChange}
      />
      <div className="card text-white bg-primary mb-3">
        <div className="card-body">
          <h3 className="card-title">
            {portfolio.netWorth && (
              <span>
                {NUMBER_UTILS.convertNumToMoney(
                  portfolio.netWorth.totalNetWorth,
                  portfolio.currencyCode,
                  useRouter().locale
                )}
              </span>
            )}
          </h3>
          <span className="card-text">
            <i className="fas fa-balance-scale mr-2"></i>
            {t("totalNetWorth")}
          </span>
        </div>
      </div>
      <AssetCategoryList
        currency={portfolio.currencyCode}
        categories={portfolio.lineItems.assets}
        onChange={handleAssetChange}
        disabled={isLoading}
        total={portfolio.netWorth ? portfolio.netWorth.totalAssets : undefined}
      />
      <LiabilityCategoryList
        currency={portfolio.currencyCode}
        categories={portfolio.lineItems.liabilities}
        onChange={handleLiabilitiesChange}
        disabled={isLoading}
        total={
          portfolio.netWorth ? portfolio.netWorth.totalLiabilities : undefined
        }
      />
    </>
  );
};

const NetWorthContainer = withTranslation()(Wrapper);

export { NetWorthContainer, Wrapper };
