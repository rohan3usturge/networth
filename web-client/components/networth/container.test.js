import { shallow } from "enzyme";
import { Wrapper } from "./container";

jest.mock("next/router", () => {
  return {
    useRouter: () => ({ locale: "en-ca" }),
  };
});

describe("NetWorthContainer", () => {
  const portfolio = {
    netWorth: {},
    currencyCode: "CAD",
    lineItems: { assets: [], liabilities: [] },
  };
  it("should render properly and match snapshot", () => {
    const wrapper = shallow(
      <Wrapper defaultPortfolio={portfolio} t={(key) => key} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
