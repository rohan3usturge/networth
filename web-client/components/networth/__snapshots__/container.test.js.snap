// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetWorthContainer should render properly and match snapshot 1`] = `
<Fragment>
  <withI18nextTranslation(Wrapper)
    onChange={[Function]}
    selectedCurrency="CAD"
  />
  <div
    className="card text-white bg-primary mb-3"
  >
    <div
      className="card-body"
    >
      <h3
        className="card-title"
      >
        <span />
      </h3>
      <span
        className="card-text"
      >
        <i
          className="fas fa-balance-scale mr-2"
        />
        totalNetWorth
      </span>
    </div>
  </div>
  <withI18nextTranslation(Wrapper)
    categories={Array []}
    currency="CAD"
    disabled={false}
    onChange={[Function]}
  />
  <withI18nextTranslation(Wrapper)
    categories={Array []}
    currency="CAD"
    disabled={false}
    onChange={[Function]}
  />
</Fragment>
`;
